Ability to build a bicep file from VSCode file context menu

bhsubra opened the pull request:
Fixes #2868 

- Added option to build bicep file from file context menu
- Created new output channel- Build to log messages regarding success/failure
- Added keybinding - Ctrl+Shift+B

![Bicep_Build_4](https://user-images.githubusercontent.com/30270536/127549706-cd784ebf-d263-4351-9bea-224963325a8b.gif)

MarcusFelling commented:
Can we display warning and errors in the Output window?

NIT: can we change "created _compiled_ file" to "created _transpiled ARM template_" or maybe simplify it to just "created file"?

alex-frankel commented:
Cool!

Is "Build" a generic Output channel or is it created for this specific scenario in Bicep? I'm wondering if it can be something more specific to Bicep (i.e. "Bicep Build" or "Bicep Operations"). Is this the same Output channel we would use in a deploy scenario? I know those are on the short list of future scenarios for this sort of functionality.

MarcusFelling commented:
> Cool!
> 
> Is "Build" a generic Output channel or is it created for this specific scenario in Bicep? I'm wondering if it can be something more specific to Bicep (i.e. "Bicep Build" or "Bicep Operations"). Is this the same Output channel we would use in a deploy scenario? I know those are on the short list of future scenarios for this sort of functionality.

@alex-frankel Good point. Bhavya was able to create a Bicep Operations channel. I'm unsure about the details on the deploy scenario. I'm working with the Azure Tools team to understand what our options are.

MarcusFelling commented:
@anthony-c-martin @majastrz @alex-frankel what do you guys think about suppressing the generated ARM Template JSON file? Ideally, customers should not have to work with this file. If they need it in their pipelines to run things like arm-ttk, they can use the cli, but the VS Code experience should abstract that need.

bhsubra commented:
> Can we display warning and errors in the Output window?
> 
> NIT: can we change "created _compiled_ file" to "created _transpiled ARM template_" or maybe simplify it to just "created file"?

Updated to display errors/warnings in output window. Thanks!

majastrz commented:
> @anthony-c-martin Anthony Martin FTE @majastrz Marcin Jastrzebski FTE @alex-frankel Alex Frankel FTE what do you guys think about suppressing the generated ARM Template JSON file? Ideally, customers should not have to work with this file. If they need it in their pipelines to run things like arm-ttk, they can use the cli, but the VS Code experience should abstract that need.

What do you mean by "suppressing the generated ARM Template JSON file"?

majastrz commented:
> Can we display warning and errors in the Output window?
> 
> NIT: can we change "created _compiled_ file" to "created _transpiled ARM template_" or maybe simplify it to just "created file"?

I like "created file".

majastrz commented on src/Bicep.LangServer/Handlers/BicepBuildCommandHandler.cs:
```diff
@@ -0,0 +1,125 @@
+                IDiagnostic diagnostic = diagnostics.ElementAt(i);
+                var (startLine, startChar) = TextCoordinateConverter.GetPosition(lineStarts, diagnostic.Span.Position);
+
+                sb.AppendLine((i+1) + ". " + diagnostic.Message + " bicep(" + diagnostic.Code + "). " + "[" + (startLine + 1) + ", " + (startChar + 1) + "]");
```
>AppendLine [](http://example.com/codeflow?start=19&length=10)

Can we try the same exact format that the CLI returns? It's a standard format used by tsc, msbuild, and other tools like that. 

VS code runs "problem matchers" (not sure what the official name is) on the output and turns errors into clickable links. I'm wondering if we could get the same behavior by matching the output format.

bhsubra replied:
Thanks for the feedback. Didn't know about the problem matchers. Updated here:
https://github.com/Azure/bicep/commit/afb789d670fba3182e8f077c1da6e8c5cd5483fd

majastrz commented on src/Bicep.LangServer/Handlers/BicepBuildCommandHandler.cs:
```diff
@@ -0,0 +1,125 @@
+            }
+        }
+
+        private string? GetCompiledFilePath(string bicepFilePath, StringBuilder sb, out string bicepFile, out string compiledFile)
```
>GetCompiledFilePath [](http://example.com/codeflow?start=24&length=19)

We have a `PathHelper` class that implements similar logic for the CLI. Can we reuse that?

bhsubra replied:
Updated as part of this commit:
https://github.com/Azure/bicep/commit/daa24e89c632c399b92f0de41c98fcea77b930eb

majastrz commented on src/Bicep.LangServer/Handlers/BicepBuildCommandHandler.cs:
```diff
@@ -0,0 +1,125 @@
+            }
+
+            sb.AppendLine("Build started...");
+            sb.AppendLine("bicep build " + bicepFile);
```
>bicep build [](http://example.com/codeflow?start=27&length=11)

We should not log bicep build here because we're not actually calling the Bicep CLI.

majastrz replied:
I also wouldn't have expected a `GetCompiledFilePath` method to have a logging side effect.

bhsubra replied:
Removed the logging and GetCompiledFilePath. Using PathHelper instead:
https://github.com/Azure/bicep/commit/daa24e89c632c399b92f0de41c98fcea77b930eb

majastrz commented on src/Bicep.LangServer/Handlers/BicepBuildCommandHandler.cs:
```diff
@@ -0,0 +1,125 @@
+                return null;
+            }
+
+            sb.AppendLine("Build started...");
```
>started [](http://example.com/codeflow?start=33&length=7)

The contents of the string builder are returned all at once back to the LSP client, so the user will never see it show up with the diagnostics and the final messages. Is there any value to logging it?

bhsubra replied:
Removed the logging:
https://github.com/Azure/bicep/commit/daa24e89c632c399b92f0de41c98fcea77b930eb

majastrz commented on src/Bicep.LangServer/Handlers/BicepBuildCommandHandler.cs:
```diff
@@ -0,0 +1,125 @@
+                return sb.ToString();
+            }
+
+            using (FileStream fileStream = new FileStream(compiledFilePath, FileMode.Create, FileAccess.ReadWrite, FileShare.ReadWrite))
```
>FileStream [](http://example.com/codeflow?start=47&length=10)

A couple of thoughts here:
1. If the user has a JSON file with the same name at the location that wasn't generated by Bicep before, we should fail the Build command. Otherwise, we could wipe out the user's data. (A workspace may not be under version control). This is the first write operation we're going to do in the user's workspace that can't be undone with Ctrl+Z like codefixes can. 

I think we have to be careful with this. The best option is probably to check if the template contains Bicep _generator metadata. It's not perfect, but should cover the most cases. (The library that can deserialize templates is already available in Bicep.)

2. `FileShare.ReadWrite` means that some other process is allowed to read the file while we're writing, which probably won't do anything useful. We should probably just use the default `FileShare` setting for now.

miqm replied:
could we have just a popup saying if user wants to override existing file? Usually we'd use build to preview the arm template and this command would be used by more advanced users that wouldn't mind hitting enter one more time.

More user-friendly command would be ability to deploy template to Azure directly but that's another issue.

majastrz replied:
The typical workflow for the feature will be make a change, click build, make a change, click build. It could get pretty annoying to have to click yes every time if someone does that after every change. However, if they only click Build once in a while, a confirmation prompt makes sense. It is also the safest option because it's user's responsibility to protect their own files and make correct choices. It's also much simpler to implement.

bhsubra replied:
Added check to see if file already exists and fail the build as part of this commit:
https://github.com/Azure/bicep/commit/daa24e89c632c399b92f0de41c98fcea77b930eb

Let me know if I am missing something

bhsubra replied:
@majastrz , as discussed offline, added logic to replace file only if template contains bicep generator metadata:
https://github.com/Azure/bicep/commit/e7e05ab2faec4b5977345b5f5d5e672d4d59d28e

Let me know if I am missing something.

majastrz commented on src/Bicep.LangServer/Handlers/BicepBuildCommandHandler.cs:
```diff
@@ -0,0 +1,125 @@
+                return sb.ToString();
+            }
+
+            using (FileStream fileStream = new FileStream(compiledFilePath, FileMode.Create, FileAccess.ReadWrite, FileShare.ReadWrite))
```
>FileShare.ReadWrite [](http://example.com/codeflow?start=115&length=19)

`FileShare.ReadWrite` means that some other process is allowed to read the file while we're writing, which probably won't do anything useful. We should probably just use the default `FileShare` setting for now.

bhsubra replied:
Updated as part of this commit:
https://github.com/Azure/bicep/commit/daa24e89c632c399b92f0de41c98fcea77b930eb

majastrz commented on src/Bicep.LangServer.UnitTests/BicepCompilationManagerTests.cs:
```diff
@@ -396,9 +391,8 @@ public void NormalUpsertAfterFatalException_ShouldReplaceDiagnostics(string? lan
-            var document = CreateMockDocument(p => receivedParams = p);
-
-            var server = CreateMockServer(document);
+            var document = BicepCompilationManagerHelper.CreateMockDocument(p => receivedParams = p);
```
>BicepCompilationManagerHelper [](http://example.com/codeflow?start=27&length=29)

Thanks for refactoring!

MarcusFelling commented:
> > @anthony-c-martin Anthony Martin FTE @majastrz Marcin Jastrzebski FTE @alex-frankel Alex Frankel FTE what do you guys think about suppressing the generated ARM Template JSON file? Ideally, customers should not have to work with this file. If they need it in their pipelines to run things like arm-ttk, they can use the cli, but the VS Code experience should abstract that need.
> 
> What do you mean by "suppressing the generated ARM Template JSON file"?

@majastrz I mean not outputting the file to the user's workspace.

alex-frankel commented:
> > > @anthony-c-martin Anthony Martin FTE @majastrz Marcin Jastrzebski FTE @alex-frankel Alex Frankel FTE what do you guys think about suppressing the generated ARM Template JSON file? Ideally, customers should not have to work with this file. If they need it in their pipelines to run things like arm-ttk, they can use the cli, but the VS Code experience should abstract that need.
> > 
> > 
> > What do you mean by "suppressing the generated ARM Template JSON file"?
> 
> @majastrz I mean not outputting the file to the user's workspace.

`bicep build` today means generate the JSON file and I don't think we should change that. Potentially we could have a different command for "check the file for errors and warnings without generating a file" -- maybe `bicep validate` or `bicep lint` (though the latter doesn't capture that non linter things will be checked). That sounds like what you are going for @marcusfelling?

MarcusFelling commented:
> > > > @anthony-c-martin Anthony Martin FTE @majastrz Marcin Jastrzebski FTE @alex-frankel Alex Frankel FTE what do you guys think about suppressing the generated ARM Template JSON file? Ideally, customers should not have to work with this file. If they need it in their pipelines to run things like arm-ttk, they can use the cli, but the VS Code experience should abstract that need.
> > > 
> > > 
> > > What do you mean by "suppressing the generated ARM Template JSON file"?
> > 
> > 
> > @majastrz I mean not outputting the file to the user's workspace.
> 
> `bicep build` today means generate the JSON file and I don't think we should change that. Potentially we could have a different command for "check the file for errors and warnings without generating a file" -- maybe `bicep validate` or `bicep lint` (though the latter doesn't capture that non linter things will be checked). That sounds like what you are going for @MarcusFelling?

Yes `bicep validate` is what I was hoping `bicep build` would evolve to.

majastrz commented:
+1 on `bicep validate` instead of adding a switch to `bicep build`. The output from `bicep validate` should be identical to the output of `bicep build` except for writing out the JSON file.

miqm commented:
> +1 on `bicep validate` instead of adding a switch to `bicep build`. The output from `bicep validate` should be identical to the output of `bicep build` except for writing out the JSON file.

Aren't we doing this on every key stroke in vs? In general I like the idea of cli command but I'd like to have ability to view json in vs (perhaps as a preview only without generating json file)

majastrz commented:
> > +1 on `bicep validate` instead of adding a switch to `bicep build`. The output from `bicep validate` should be identical to the output of `bicep build` except for writing out the JSON file.
> 
> Aren't we doing this on every key stroke in vs? In general I like the idea of cli command but I'd like to have ability to view json in vs (perhaps as a preview only without generating json file)

Yeah, I think this is just about adding the verb to the CLI. We're good on VS code side.

MarcusFelling commented:
@bhsubra @majastrz @anthony-c-martin Is there any way to get colored output in the output window? Without needing to install an extension like this? [https://marketplace.visualstudio.com/items?itemName=IBM.output-colorizer](https://marketplace.visualstudio.com/items?itemName=IBM.output-colorizer)

bhsubra commented:
> @bhsubra @majastrz @anthony-c-martin Is there any way to get colored output in the output window? Without needing to install an extension like this? https://marketplace.visualstudio.com/items?itemName=IBM.output-colorizer

I looked into this a bit last night and couldn't find any documentation supporting colorized output in output window. I found this proposal, but looks like support is available for tasks
https://github.com/Microsoft/vscode/issues/1801

@majastrz , @anthony-c-martin, let me know if you have any suggestions. Thanks!

majastrz commented:
The detection of `_generator` works great, but I think we need to improve the error message:
![image](https://user-images.githubusercontent.com/22460039/127407080-cb091709-4ab3-4ba9-9486-44bfbeb11410.png)

I thinking something like this:
Build failed. The file "main.json" already exists and was not generated by Bicep. If overwriting the file is intended, delete it manually and retry the Build command.

majastrz commented on src/Bicep.LangServer/Handlers/BicepBuildCommandHandler.cs:
```diff
@@ -0,0 +1,130 @@
+        {
+            if (!string.IsNullOrEmpty(template))
+            {
+                JToken jtoken = template.FromJson<JToken>();
```
>FromJson [](http://example.com/codeflow?start=41&length=8)

This will throw when JSON is malformed.

bhsubra replied:
Added error handling here:
https://github.com/Azure/bicep/commit/437f01c6f30eb6074c372ff7e91b31b0b7c678e7

majastrz commented:
When I have a malformed main.json file next to a main.bicep file, clicking Build doesn't do anything. (I was trying in GitHub codespaces, so the behavior might be slightly different locally.) I suspect it's because it's throwing internally due to the JSON not deserializing. I added a comment in the code as well...

majastrz commented:
> > @bhsubra
> >             
> >             Bhavya Subramanian
> >             FTE @majastrz
> >             
> >             Marcin Jastrzebski
> >             FTE @anthony-c-martin
> >             
> >             Anthony Martin
> >             FTE Is there any way to get colored output in the output window? Without needing to install an extension like this? https://marketplace.visualstudio.com/items?itemName=IBM.output-colorizer
> 
> I looked into this a bit last night and couldn't find any documentation supporting colorized output in output window. I found this proposal, but looks like support is available for tasks
> [microsoft/vscode#1801](https://github.com/microsoft/vscode/issues/1801)
> 
> @majastrz Marcin Jastrzebski FTE , @anthony-c-martin Anthony Martin FTE, let me know if you have any suggestions. Thanks!

Makes sense. I don't think I've ever seen colorized output there. I suspect it's possible to do by sending terminal control codes, but not sure how fragile it would be. This isn't really the primary troubleshooting experience (the user should be using VS code problems to figure out why things aren't valid), so I think what you have is pretty good:
![image](https://user-images.githubusercontent.com/22460039/127408242-5fbab39b-4fa9-4e9b-b4a8-6446885da4da.png)

majastrz commented on src/vscode-bicep/src/commands/build.ts:
```diff
@@ -0,0 +1,40 @@
+      }
+    );
+
+    appendToOutputChannel(this.outputChannel, buildOutput);
```
>appendToOutputChannel [](http://example.com/codeflow?start=4&length=21)

Btw, do we need to do something here so unhandled exceptions in the command get surfaced correctly in VS code?

My malformed JSON case should have thrown an exception in the language server, but nothing happened client side anywhere. When this happens with other LSP requests, you get a popup that the request failed.

bhsubra replied:
Added error handling logic here:
https://github.com/Azure/bicep/pull/3708#issuecomment-888676631

Added special case for malformed json to output below message:
![image](https://user-images.githubusercontent.com/30270536/127523187-8762aca4-5378-406d-a8df-697913f62a19.png)

majastrz replied:
I think the popup gets tracked with telemetry, but failures in the output channel are not. I think we should stay consistent with the behavior for all other requests.

bhsubra replied:
As discussed offline, invoking client.error(..) to show popup.

Now that we are returning false for malformed json, I couldn't think of a good way to verify this. So I set context to null in debugger and verified popup shows up and redirects to bicep output channel. Here's the gif:

![Bicep_Build_5](https://user-images.githubusercontent.com/30270536/127549564-5e7912f3-5cd4-423e-9778-4d42e1789633.gif)

bhsubra commented:
> and was not generated by Bicep. If overwriting the file is intended, delete it manually and retry the Build command.

Updated here:
https://github.com/Azure/bicep/pull/3708#issuecomment-888676631

Thanks!

bhsubra commented:
> When I have a malformed main.json file next to a main.bicep file, clicking Build doesn't do anything. (I was trying in GitHub codespaces, so the behavior might be slightly different locally.) I suspect it's because it's throwing internally due to the JSON not deserializing. I added a comment in the code as well...

Added error handling here:
https://github.com/Azure/bicep/commit/437f01c6f30eb6074c372ff7e91b31b0b7c678e7

majastrz commented on src/Bicep.LangServer/Handlers/BicepBuildCommandHandler.cs:
```diff
@@ -0,0 +1,138 @@
+                    return "Build failed. The file \"" + compiledFile + "\" already exists and was not generated by Bicep. If overwriting the file is intended, delete it manually and retry the Build command.";
+                }
+            }
+            catch (JsonReaderException)
```
>JsonReaderException [](http://example.com/codeflow?start=19&length=19)

Are we sure that the template library can't throw any other exceptions when the template is malformed? It seems like it'll be safer to just catch `Exception` and return false in `TemplateContainsBicepGeneratorMetadata()`. 

(We don't really need to give the user the reason why we think Bicep didn't generate the file.)

bhsubra replied:
As discussed offline, updated to catch exception and return false:
https://github.com/Azure/bicep/commit/79e7a2b9d06e87589572453b62ebb06cbaaab30f

majastrz reviewed (approved):
LGTM!
